### PR TITLE
Draft: AppConfig integration

### LINKED ISSUE #465: AppConfig support
**Type**: Feature

**Is your feature request related to a problem? Please describe.**
I would like to fetch configuration from [AppConfig](https://docs.aws.amazon.com/appconfig/latest/userguide/what-is-appconfig.html) on application start up. This way I can keep my whole configuration away from my code and keep it in a same place as configuration for other applications.

**Describe the solution you'd like**
For this I would like to use `Spring.config.import`. 
Example:
`spring.config.import: aws-appconfig:applicationId___configurationId___environment`
We can use another char for splitting `___` 3 * _ is just an example.

**Describe alternatives you've considered**
Currently I would have to write my own implementation

**Additional context**
https://sdk.amazonaws.com/java/api/latest/software/amazon/awssdk/services/appconfigdata/AppConfigDataClient.html

### PR DESCRIPTION
**Pull request by MatejNedic**
## :loudspeaker: Type of change

- [ ] Bugfix
- [x] New feature
- [ ] Enhancement
- [ ] Refactoring

## :scroll: Description
Still have to go through it once more time and add some docs. Tried against real AWS and all tests are passing.
Since AWS Localstack has a bug it is not possible to make a real integration test.
Closes #465 

@eddumelendez SInce you worked on AppConfig integration before please when you have time check the design. I will do more cleaning and polishing of the code.

## :pencil: Checklist

- [x] I reviewed submitted code
- [x] I added tests to verify changes
- [x] I updated reference documentation to reflect the change
- [x] All tests passing
- [x] No breaking changes

### CODE REVIEW

File: spring-cloud-aws-autoconfigure/pom.xml [hunk 1]
```diff
@@ -60,6 +60,16 @@
...
 			<optional>true</optional>
 		</dependency>
+		<dependency>
+			<groupId>software.amazon.awssdk</groupId>
+			<artifactId>appconfigdata</artifactId>
+			<optional>true</optional>
+		</dependency>
+		<dependency>
+		<groupId>io.awspring.cloud</groupId>
+		<artifactId>spring-cloud-aws-app-config</artifactId>
```
Comment (MatejNedic) [inline spring-cloud-aws-autoconfigure/pom.xml:70]: formatting is wrong here not sure why spotless did not format it.

File: spring-cloud-aws-autoconfigure/src/main/java/io/awspring/cloud/autoconfigure/config/appconfig/AppConfigDataLocationResolver.java [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+					"No AppConfig keys provided in `spring.config.import=aws-appconfig:` configuration.");
+		}
+		return locations;
+	}
+
+	public static RequestContext splitContext(String context) {
+		RequestContext.Builder builder = RequestContext.builder();
+		AtomicReference<Integer> i = new AtomicReference<>(0);
+		builder.context(context);
+		Arrays.stream(context.split("___")).forEach(split -> {
```
Comment (MatejNedic) [inline spring-cloud-aws-autoconfigure/src/main/java/io/awspring/cloud/autoconfigure/config/appconfig/AppConfigDataLocationResolver.java:84]: Should the prefix that is splitting values be adjustable or changed to something else? 
I think even a simple `/` could do but I am afraid there will be cases where someone has `/` in ConfigurationProfileName or ApplicationName

File: spring-cloud-aws-autoconfigure/src/main/resources/META-INF/spring.factories [hunk 3]
```diff
@@ -12,14 +12,17 @@ io.awspring.cloud.autoconfigure.dynamodb.DynamoDbAutoConfiguration
 # ConfigData Location Resolvers
 org.springframework.boot.context.config.ConfigDataLocationResolver=\
 io.awspring.cloud.autoconfigure.config.parameterstore.ParameterStoreConfigDataLocationResolver,\
+io.awspring.cloud.autoconfigure.config.appconfig.AppConfigDataLocationResolver,\
```
Comment (eddumelendez) [inline spring-cloud-aws-autoconfigure/src/main/resources/META-INF/spring.factories:15]: order should be alphabetically. it applies for the other two entries

File: spring-cloud-aws-autoconfigure/src/main/java/io/awspring/cloud/autoconfigure/config/appconfig/AppConfigDataLocationResolver.java [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+		if (!location.isOptional() && locations.isEmpty()) {
+			throw new AppConfigKeysMissingException(
+					"No AppConfig keys provided in `spring.config.import=aws-appconfig:` configuration.");
+		}
+		return locations;
+	}
+
+	public static RequestContext splitContext(String context) {
+		RequestContext.Builder builder = RequestContext.builder();
+		AtomicReference<Integer> i = new AtomicReference<>(0);
```
Comment (eddumelendez) [inline spring-cloud-aws-autoconfigure/src/main/java/io/awspring/cloud/autoconfigure/config/appconfig/AppConfigDataLocationResolver.java:86]: any specific use? maybe I'm missing something

File: spring-cloud-aws-autoconfigure/src/main/java/io/awspring/cloud/autoconfigure/config/appconfig/AppConfigDataLocationResolver.java (hunk 2 above)
Comment (eddumelendez) [reply spring-cloud-aws-autoconfigure/src/main/java/io/awspring/cloud/autoconfigure/config/appconfig/AppConfigDataLocationResolver.java:88]: I would go with the simple delimiter `/` and wait for feedback on the next milestone. Given you can have any character in the names. Also, for better debugging trace for loop would be enough

File: spring-cloud-aws-autoconfigure/src/test/java/io/awspring/cloud/autoconfigure/config/appconfig/AppConfigDataLoaderTest.java [hunk 5]
```diff
@@ -0,0 +1,215 @@
...
+import software.amazon.awssdk.http.SdkHttpClient;
+import software.amazon.awssdk.http.apache.ApacheHttpClient;
+import software.amazon.awssdk.services.appconfigdata.AppConfigDataClient;
+import software.amazon.awssdk.services.appconfigdata.model.GetLatestConfigurationRequest;
+import software.amazon.awssdk.services.appconfigdata.model.GetLatestConfigurationResponse;
+import software.amazon.awssdk.services.appconfigdata.model.StartConfigurationSessionRequest;
+import software.amazon.awssdk.services.appconfigdata.model.StartConfigurationSessionResponse;
+
+@ExtendWith(OutputCaptureExtension.class)
+class AppConfigDataLoaderTest {
```
Comment (eddumelendez) [inline spring-cloud-aws-autoconfigure/src/test/java/io/awspring/cloud/autoconfigure/config/appconfig/AppConfigDataLoaderTest.java:51]: not specific to this PR but wondering that given the project relies on integration with AWS APIs, tests should also run on top of localstack in this part of the project?

File: spring-cloud-aws-app-config/src/main/java/io/awspring/cloud/appconfig/RequestContext.java [hunk 6]
```diff
@@ -0,0 +1,115 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.awspring.cloud.appconfig;
+
+/**
+ * @author Matej Nedic
+ * @since 3.0
+ */
+public class RequestContext {
```
Comment (eddumelendez) [inline spring-cloud-aws-app-config/src/main/java/io/awspring/cloud/appconfig/RequestContext.java:22]: From AppConfig perspective, there is known attributes such as `configurationProfileIdentifier`, `environmentIdentifier` and `applicationIdentifier`.  However, `context` is something to be used internally AFAICS. I would suggest to use and internal format to be built once the other attributes have been filled.

**Review by eddumelendez — COMMENTED** [review]
My concern is related to the token management

From [docs](https://docs.aws.amazon.com/appconfig/2019-10-09/APIReference/API_appconfigdata_StartConfigurationSession.html#API_appconfigdata_StartConfigurationSession_ResponseSyntax)

> This token should only be used once in your first call to GetLatestConfiguration. You must use the new token in the GetLatestConfiguration response (NextPollConfigurationToken) in each subsequent call to GetLatestConfiguration.
>
> This token and the NextPollConfigurationToken token expire after 24 hours. If a GetLatestConfiguration call uses an expired token, the system returns BadRequestException.

Also, regarding to 

> Configuration
The data of the configuration. This may be empty if the client already has the latest version of configuration.

is the integration already managing this? I guess the token is the one that is used here and in every restart the app will generate a new token and will perform those request to fetch the latest configuration. However, what about refresh? I think the result is the same but just writing down some doubts I have regarding to this.

Source: https://docs.aws.amazon.com/appconfig/2019-10-09/APIReference/API_Operations_AWS_AppConfig_Data.html

File: spring-cloud-aws-autoconfigure/src/test/java/io/awspring/cloud/autoconfigure/config/appconfig/AppConfigDataLoaderTest.java (hunk 5 above)
Comment (MatejNedic) [reply spring-cloud-aws-autoconfigure/src/test/java/io/awspring/cloud/autoconfigure/config/appconfig/AppConfigDataLoaderTest.java:51]: Sadly AppconfigData is not supported by [Localstack ](https://localstack.cloud/features/).
I will open an Issue on AWS Loclastack GitHub about it.

File: spring-cloud-aws-app-config/src/main/java/io/awspring/cloud/appconfig/RequestContext.java (hunk 6 above)
Comment (MatejNedic) [reply spring-cloud-aws-app-config/src/main/java/io/awspring/cloud/appconfig/RequestContext.java:22]: So we should pass `String context` which is full path and in `AppConfigPropertySource` split it up to `configurationProfileIdentifier`, `environmentIdentifier` and `applicationIdentifier`? If I understood it right.

File: spring-cloud-aws-autoconfigure/src/test/java/io/awspring/cloud/autoconfigure/config/appconfig/AppConfigDataLoaderTest.java (hunk 5 above)
Comment (willome) [reply spring-cloud-aws-autoconfigure/src/test/java/io/awspring/cloud/autoconfigure/config/appconfig/AppConfigDataLoaderTest.java:50]: AppConfig is supported on Localstack Pro. See https://docs.localstack.cloud/user-guide/aws/appconfig/

**Comment by noshua** [comment]
What is holding this back? I'm waiting eagerly for this feature since months.

File: spring-cloud-aws-autoconfigure/src/test/java/io/awspring/cloud/autoconfigure/config/appconfig/AppConfigDataLoaderTest.java [hunk 7]
```diff
@@ -0,0 +1,381 @@
...
+	@BeforeAll
+	public static void createApplication() {
+		AppConfigClient appConfigClient = AppConfigClient.builder()
+			.endpointOverride(localstack.getEndpointOverride(DYNAMODB)).region(Region.of(localstack.getRegion()))
+			.credentialsProvider(StaticCredentialsProvider
+				.create(AwsBasicCredentials.create(localstack.getAccessKey(), localstack.getSecretKey())))
+			.build();
+		String appId = appConfigClient.createApplication(CreateApplicationRequest.builder().name("app").description("desc").build()).id();
+		String envId = appConfigClient.createEnvironment(CreateEnvironmentRequest.builder().applicationId(appId).description("desc").name("env").build()).id();
+		appConfigClient.createDeploymentStrategy()
```
Comment (maciejwalkowiak) [inline spring-cloud-aws-autoconfigure/src/test/java/io/awspring/cloud/autoconfigure/config/appconfig/AppConfigDataLoaderTest.java:87]: this looks like its not finished?

File: spring-cloud-aws-autoconfigure/src/test/java/io/awspring/cloud/autoconfigure/config/appconfig/AppConfigDataLoaderTest.java (hunk 5 above)
Comment (MatejNedic) [reply spring-cloud-aws-autoconfigure/src/test/java/io/awspring/cloud/autoconfigure/config/appconfig/AppConfigDataLoaderTest.java:69]: https://github.com/localstack/localstack/issues/6891
Sadly not what we need.

**Comment by MatejNedic** [comment]
It would be cool to include tests with localstack once is implemented ->
https://github.com/localstack/localstack/issues/6891

**Comment by yurybubnov** [comment]
I'm also eagerly waiting for the feature. How can I help to get it integrated?

**Comment by wmz7year** [comment]
any updates?

**Comment by mikelhamer** [comment]
This feature would be awesome! Keeping an eye on this, and willing to help if needed.

**Comment by NChitty** [comment]
I feel like I must have missed it but can you clarify what happens when a new AppConfig `HostedVersion` gets pushed to the application and environment? Does it force a reload of the context?

**Comment by maciejwalkowiak** [comment]
@MatejNedic Localstack implemented Appconfig support

**Comment by MatejNedic** [comment]
@maciejwalkowiak didn't know that perfect. Will implement with LocalStack tests and we can then hopefully merge and release.
One more thing is that we will need to put LocalStack PRO credentials in GitHub pipeline

**Comment by MatejNedic** [comment]
Hey @NChitty , if you enable reload inside of integration it will periodically call AppConfig and do a reload or refresh depending on what you choose. It is covered by documentation.

**Comment by noshua** [comment]
I don't wanna be rude, but over a year has passed by since my last status request. Is this feature dead before arrival?!

**Comment by MatejNedic** [comment]
Hey @noshua , I am thinking about redoing this integration. I have to rethink it, don't like current implementation can do more things will be exploring it.

**Pull request closed without merging**